crates/ruma-api/src/lib.rs

```diff
@@ -421,3 +425,129 @@ pub struct Metadata {
     /// What authentication scheme the server uses for this endpoint.
     pub authentication: AuthScheme,
 }
+
+/// The matrix versions ruma currently understands to exist.
+///
+/// Matrix, since fall 2021, has a quarterly release schedule, using a global `vX.Y` versioning
+/// scheme.
+///
+/// Every new minor version denotes stable support for endpoints in a *relatively*
+/// backwards-compatible manner.
+///
+/// Matrix has a deprecation policy, read more about it here: <https://spec.matrix.org/v1.2/#deprecation-policy>.
+// TODO add the following once `EndpointPath` and added/deprecated/removed macros are in place;
+// Ruma keeps track of when endpoints are added, deprecated, and removed. It'll automatically
+// select the right endpoint stability variation to use depending on which matrix version you pass
+// it with [`EndpointPath`], see its respective documentation for more.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum MatrixVersion {
+    /// Version 1.0 of the matrix specification
+    ///
+    /// Retroactively defined as https://spec.matrix.org/v1.1/#legacy-versioning
+    V1_0,
+
+    /// Version 1.1 of the matrix specification.
+    ///
+    /// See https://spec.matrix.org/v1.1/
+    V1_1,
+
+    /// Version 1.2 of the matrix specification.
+    ///
+    /// See https://spec.matrix.org/v1.2/
+    V1_2,
+}
+
+/// An error that happens when ruma cannot understand a matrix version.
+#[derive(Debug)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct UnknownVersionError;
+
+impl Display for UnknownVersionError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Version string was unknown.")
+    }
+}
+
+impl StdError for UnknownVersionError {}
+
+impl TryInto<MatrixVersion> for &str {
```

Comment:
```suggestion
impl TryFrom<&str> for MatrixVersion {
```

Reply:
I'll have to change the entire impl around a bit, but understood :+1: